test/varmat_compatibility.py

```diff
@@ -0,0 +1,260 @@
+#!/usr/bin/python
+
+import itertools
+import json
+from argparse import ArgumentDefaultsHelpFormatter, ArgumentParser
+import os
+import Queue
+import subprocess
+import sys
+import tempfile
+import threading
+
+from sig_utils import make, handle_function_list, get_signatures
+from signature_parser import SignatureParser
+from code_generator import CodeGenerator
+
+HERE = os.path.dirname(os.path.realpath(__file__))
+TEST_FOLDER = os.path.abspath(os.path.join(HERE, "..", "test"))
+sys.path.append(TEST_FOLDER)
+WORKING_FOLDER = "test/varmat-compatibility"
+
+BENCHMARK_TEMPLATE = """
+static void {benchmark_name}() {{
+{code}
+}}
+"""
+
+def run_command(command):
+    """
+    Runs given command and waits until it finishes executing.
+    :param command: command to execute
+    """
+    proc = subprocess.Popen(command, stdout = subprocess.PIPE, stderr = subprocess.PIPE)
+    stdout, stderr = proc.communicate()
+    
+    if proc.poll() == 0:
+        return (True, stdout, stderr)
+    else:
+        return (False, stdout, stderr)
+
+def build_signature(cpp_code):
+    """
+    Try to build the given cpp code
+
+    Return a tuple (successful, attempted)
+
+    attempted is true if there was an attempt to build the code
+
+    successful is true is the code was successfully built
+
+    :param cpp_code Code to build
+    """
+    attempted = False
+    successful = False
+
+    if cpp_code:
+        attempted = True
+
+        f = tempfile.NamedTemporaryFile("w", dir = WORKING_FOLDER, suffix = "_test.cpp", delete = False)
+        f.write("#include <test/expressions/expression_test_helpers.hpp>\n\n")
+        f.write(cpp_code)
+        f.close()
+
+        cpp_path = os.path.join(WORKING_FOLDER, os.path.basename(f.name))
+
+        object_path = cpp_path.replace(".cpp", ".o")
+        dependency_path = cpp_path.replace(".cpp", ".d")
+        stdout_path = cpp_path.replace(".cpp", ".stdout")
+        stderr_path = cpp_path.replace(".cpp", ".stderr")
```

Comment:
I put them in files so I could read them after the fact for debugging. Maybe on error I should immediately print stdout/stderr and kill the process though. That might be simpler.

Reply:
Oh oh, what's happening here is we're only storing the stdout/stderr if the check fails. So if the check fails we keep everything (so someone can go dig in `test/varmat-compatibility/` and see what happened), and if it passes everything gets deleted. It's not the best but it works.

Reply:
It's kind of like expressions -- if you have a problem go dig in `test/expressions/tests0_test.cpp`. I guess the annoying part is that the file is named badly. I added a prefix argument so that the function name can be used to prefix the temporary files to make it slightly easier to sort through.

Reply:
As I understand it, an error here just means a signature is not compatible, right? So you should not kill the whole test for this. Anyway, maybe just use a single file as a log instead of one file for each failed compilation (it might be a bit too much output for stdout).

Reply:
> So you should not kill the whole test for this

Yeah I realized that and decided against it.

> Anyway, maybe just use a single file as a log instead of one file for each failed compilation (it might be a bit too much output for stdout).

That sounds cleaner

Reply:
I decided it was simpler to just make the prefixes cleaner.

So now the signature `dot_product(row_vector, vector) => real` will make files named like: `dot_product_row_vector_vector_real_i7Actp_test` so it's not so bad to inspect them.

The other options were 1 stdout/stderr file per thread, but then it's confusing where stuff is, or one big stdout/stderr overall, which would need reconstructed at the end of everything -- separate files were just simpler.

Reply:
> or one big stdout/stderr overall, which would need reconstructed at the end of everything

Not really. You could write to the same file object with multiple threads if you use a lock. I would like to avoid generating a huge number of files, which this will do with one file per signature.

Reply:
Hmm well the cpp files are still there, and I do want to leave those separate. So if there are 10000 cpp files left over, this will leave 20000 stdouts as well. Which is pretty bad I agree (enough to make accidentally grepping in that folder annoying)

I could also add a --debug flag so that files are only written/kept if the flag is included?

Reply:
Hmm, yes you need to generate cpp files, but you could clean them up afterwards. I am still in favor of single output file, but adding a flag if also fine if you want to go in that direction.

Reply:
Added a debug flag. By default this won't leave any temporary files around.